### PR TITLE
[Merged by Bors] - Prioritise important parts of block processing

### PR DESCRIPTION
**Pull request by michaelsproul**
## Issue Addressed

Closes https://github.com/sigp/lighthouse/issues/2327

## Proposed Changes

This is an extension of some ideas I implemented while working on `tree-states`:

- Cache the indexed attestations from blocks in the `ConsensusContext`. Previously we were re-computing them 3-4 times over.
- Clean up `import_block` by splitting each part into `import_block_XXX`.
- Move some stuff off hot paths, specifically:
  - Relocate non-essential tasks that were running between receiving the payload verification status and priming the early attester cache. These tasks are moved after the cache priming:
    - Attestation observation
    - Validator monitor updates
    - Slasher updates
    - Updating the shuffling cache
  - Fork choice attestation observation now happens at the end of block verification in parallel with payload verification (this seems to save 5-10ms).
  - Payload verification now happens _before_ advancing the pre-state and writing it to disk! States were previously being written eagerly and adding ~20-30ms in front of verifying the execution payload. State catchup also sometimes takes ~500ms if we get a cache miss and need to rebuild the tree hash cache.

The remaining task that's taking substantial time (~20ms) is importing the block to fork choice. I _think_ this is because of pull-tips, and we should be able to optimise it out with a clever total active balance cache in the state (which would be computed in parallel with payload verification). I've decided to leave that for future work though. For now it can be observed via the new `beacon_block_processing_post_exec_pre_attestable_seconds` metric.

### CODE REVIEW

File: beacon_node/beacon_chain/src/beacon_chain.rs [hunk 1]
```diff
@@ -3077,62 +3187,105 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
...
+        block: BeaconBlockRef<T::EthSpec>,
+        block_root: Hash256,
+        current_epoch: Epoch,
+        current_finalized_checkpoint: Checkpoint,
+        current_eth1_finalization_data: Eth1FinalizationData,
+        parent_eth1_finalization_data: Eth1FinalizationData,
+        parent_block_slot: Slot,
+    ) {
+        // Do not write to eth1 finalization cache for blocks older than 5 epochs.
+        if block.slot().epoch(T::EthSpec::slots_per_epoch()) + 5 < current_epoch {
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/beacon_chain.rs:3298]: I changed this condition slightly to simplify the inputs to this function. Would appreciate you double-checking it @ethDreamer :pray:

File: beacon_node/beacon_chain/src/block_verification.rs [hunk 2]
```diff
@@ -1421,13 +1423,52 @@ impl<T: BeaconChainTypes> ExecutionPendingBlock<T> {
...
         }
 
+        /*
+         * Apply the block's attestations to fork choice.
+         *
+         * We're running in parallel with the payload verification at this point, so this is
+         * free real estate.
+         */
+        let current_slot = chain.slot()?;
+        let mut fork_choice = chain.canonical_head.fork_choice_write_lock();
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/block_verification.rs:1445]: This constitutes a "bounce" of the fork choice write lock, as we need to re-obtain it later to apply the block. I think this is acceptable as there's likely to be a large real-time delay between the two locking events due to this running during payload verification.

File: consensus/state_processing/src/consensus_context.rs [hunk 3]
```diff
@@ -48,7 +56,7 @@ impl<T: EthSpec> ConsensusContext<T> {
         state: &BeaconState<T>,
         spec: &ChainSpec,
     ) -> Result<u64, ContextError> {
-        self.check_slot(state.slot())?;
+        self.check_epoch(state.current_epoch())?;
```
Comment (paulhauner) [inline consensus/state_processing/src/consensus_context.rs:78]: This seems a little risky, but I suspect I'm missing something. Do you mind sharing the motivations for this?

File: beacon_node/beacon_chain/src/block_verification.rs [hunk 4]
```diff
@@ -1127,6 +1127,78 @@ impl<T: BeaconChainTypes> ExecutionPendingBlock<T> {
...
+            if is_valid_merge_transition_block {
+                validate_merge_block(&chain, block.message(), AllowOptimisticImport::Yes).await?;
+            };
+
+            // The specification declares that this should be run *inside* `per_block_processing`,
+            // however we run it here to keep `per_block_processing` pure (i.e., no calls to external
+            // servers).
+            //
+            // It is important that this function is called *after* `per_slot_processing`, since the
+            // `randao` may change.
```
Comment (paulhauner) [inline beacon_node/beacon_chain/src/block_verification.rs]: We can remove this statement, I was wrong to add it in the first place.

As discussed offline, the value returned from `get_randao_mix` cannot change due to `process_slot`. Although we might start reading from a different *index* in `state.randao_mixes`, the value that is returned will stay the same.

**Review by paulhauner — CHANGES_REQUESTED** [review]
Nice, this is great. I've had my eye on a couple of these optimisations as well! I think splitting the logic out into small functions is nice too.

Just a minor change and a question :pray:

File: beacon_node/beacon_chain/src/beacon_chain.rs (hunk 1 above)
Comment (ethDreamer) [reply beacon_node/beacon_chain/src/beacon_chain.rs:3298]: @michaelsproul LGTM! Very clean :)

File: consensus/state_processing/src/consensus_context.rs (hunk 3 above)
Comment (michaelsproul) [reply consensus/state_processing/src/consensus_context.rs:78]: No, you're right, it is a bit risky.

I had to make this change when I rearranged the block signature verifier to take the consensus context. When verifying chain segments in `signature_verify_chain_segment` we have the state from the same epoch which is sufficient for determining the shuffling at each block's slot, but does not have an exactly matching slot.

A safer solution might be to have two methods on `ConsensusContext` for getting the proposer index: one checking the slot that we use almost everywhere, and one checking the epoch that we use in the block signature verifier. Let me know if you think that would be better, or if you've got other ideas!

File: beacon_node/beacon_chain/src/beacon_chain.rs (hunk 1 above)
Comment (michaelsproul) [reply beacon_node/beacon_chain/src/beacon_chain.rs:3298]: Thanks!

File: consensus/state_processing/src/consensus_context.rs (hunk 3 above)
Comment (michaelsproul) [reply consensus/state_processing/src/consensus_context.rs:78]: Addressed in https://github.com/sigp/lighthouse/pull/3696/commits/d592fb3cc3ee3caaa5045c7e311c7f9f96af4e60, let me know if you think that's sufficient

**Comment by michaelsproul** [comment]
Added a fix for the pubkey cache being written to disk while the write lock is held: https://github.com/sigp/lighthouse/pull/3696/commits/b9fa16690f889475a458214138476f37fe3a157c. Turned out to be quite straight-forward!

This PR is running on Goerli and is ready for review

**Review by paulhauner — APPROVED** [review]
> Added a fix for the pubkey cache being written to disk while the write lock is held: https://github.com/sigp/lighthouse/commit/b9fa16690f889475a458214138476f37fe3a157c. Turned out to be quite straight-forward!

Ooh nice, I think this is a really neat way to do it! Much simpler than the unmerged implementation I had a while ago. I think that returning the KV ops is definitely the way to go.

I've had a think about the effects of persisting the KV ops at a later point during block import. The following objects/caches will have the block applied to them before the pubkey cache is persisted:

1. Fork choice: this is OK since we hold the fork choice lock until we've persisted the pubkeys and revert fork choice if the persistence fails.
1. `attester_cache`: this is OK since we will only try to use an attester cache value if the block is in fork choice. It also doesn't cache anything relevant to validators that were just added to a state.
1. `early_attester_cache`: this is a little funny, but not a blocker on this change. I'll expand more on this later.
1. `shuffling_cache`: this should be OK since a validator added in some state cannot be in the committee for that state.
1. observed attestations: also OK because a validator added in some block cannot have an attestation included in that block.
1. `validator_monitor`: OK because the validator monitor doesn't reference the pubkey cache (it actually never references the `BeaconChain` at all).
1. `slasher`: I think this is also OK since it just processes attestations from the block and a validator introduced to the val set in a block cannot have an attestation included in it.

Overall, I think the pubkey cache change is good! 👍 ✅ 

Regarding the `early_attester_cache`, I noticed that we're not clearing it when we [restore fork choice from the store](https://github.com/sigp/lighthouse/blob/6d5a2b509fac7b6ffe693866f58ba49989f946d7/beacon_node/beacon_chain/src/beacon_chain.rs#L2956-L2965)! I don't think it's a big deal, but it might make it possible for us to produce attestations we can't verify. It might be prudent to [clear](https://github.com/sigp/lighthouse/blob/6d5a2b509fac7b6ffe693866f58ba49989f946d7/beacon_node/beacon_chain/src/early_attester_cache.rs#L41-L45) the attester cache when we restore fork choice from the store. Are you open to doing that in this PR, I'm happy to do it separately if not.

Regarding [checking for the proposer index slot](https://github.com/sigp/lighthouse/pull/3696#discussion_r1025967428), your change looks good to me. I guess that the absolute safest way would be to store `(Slot, ProposerIndex)` in the `ConsensusContext`, however I'm not sure that's necessary.

I'm happy to merge this with or without my suggestions.

**Comment by michaelsproul** [comment]
Thanks for the review! I've added a `clear` for the early attester cache in 575cb4623606573a2f170aad87e04a8b20af4b49.

Excited to merge this next week?!

**Review by paulhauner — APPROVED** [review]
Let's merge it! 🎉

**Comment by michaelsproul** [comment]
Yay!

bors r+

**Comment by michaelsproul** [comment]
Ye olde flaky attestation test

bors retry

**Pull request closed without merging**